### PR TITLE
Enabling ollama to run on Intel GPUs with SYCL backend

### PR DESCRIPTION
**Pull request by zhewang1-intc**
Hi, I am submitting this pr to enable ollama to run on Intel GPUs with SYCL as the backend. This pr was [originally](https://github.com/ollama/ollama/pull/2458) started by @felipeagc who is currently unable to actively participate due to relocation.
The original pr had fallen behind the main branch, making it inconvenient for maintainers @mxyng @jmorganca @dhiltgen to review. Therefore, I rebased the latest main branch and opened this new pull request. I have verified that it works correctly on Ubuntu 22.04 with ARC 770 GPU.
While I am not very familiar with this project and I welcome any guidance and assistance from the community. Let’s work together to make ollama support Intel GPU platforms. cc:@hshen14 @kevinintel @airmeng

UPDATE: works well on windows10 + ARC 770
UPDATE: works well on oneapi-docker-image(oneapi-basekit-Ubuntu22.04) + ARC770

### CODE REVIEW

**Comment by airMeng** [comment]
@jmorganca @mxyng Could you give a review?

**Comment by semidark** [comment]
I tried this on my Intel integrated GPU. I am able to build and run llama.cpp with Intel GPU support without too much problems following this Tutorial: https://github.com/ggerganov/llama.cpp/blob/master/README-sycl.md

But the Version of llama.cpp started by Ollama fails to detect/use my GPU. 

```bash
./ollama serve
time=2024-03-24T11:21:16.096-06:00 level=INFO source=images.go:863 msg="total blobs: 6"
time=2024-03-24T11:21:16.096-06:00 level=INFO source=images.go:870 msg="total unused blobs removed: 0"
[GIN-debug] [WARNING] Creating an Engine instance with the Logger and Recovery middleware already attached.

[GIN-debug] [WARNING] Running in "debug" mode. Switch to "release" mode in production.
 - using env:	export GIN_MODE=release
 - using code:	gin.SetMode(gin.ReleaseMode)

[GIN-debug] POST   /api/pull                 --> github.com/jmorganca/ollama/server.PullModelHandler (5 handlers)
[GIN-debug] POST   /api/generate             --> github.com/jmorganca/ollama/server.GenerateHandler (5 handlers)
[GIN-debug] POST   /api/chat                 --> github.com/jmorganca/ollama/server.ChatHandler (5 handlers)
[GIN-debug] POST   /api/embeddings           --> github.com/jmorganca/ollama/server.EmbeddingHandler (5 handlers)
[GIN-debug] POST   /api/create               --> github.com/jmorganca/ollama/server.CreateModelHandler (5 handlers)
[GIN-debug] POST   /api/push                 --> github.com/jmorganca/ollama/server.PushModelHandler (5 handlers)
[GIN-debug] POST   /api/copy                 --> github.com/jmorganca/ollama/server.CopyModelHandler (5 handlers)
[GIN-debug] DELETE /api/delete               --> github.com/jmorganca/ollama/server.DeleteModelHandler (5 handlers)
[GIN-debug] POST   /api/show                 --> github.com/jmorganca/ollama/server.ShowModelHandler (5 handlers)
[GIN-debug] POST   /api/blobs/:digest        --> github.com/jmorganca/ollama/server.CreateBlobHandler (5 handlers)
[GIN-debug] HEAD   /api/blobs/:digest        --> github.com/jmorganca/ollama/server.HeadBlobHandler (5 handlers)
[GIN-debug] POST   /v1/chat/completions      --> github.com/jmorganca/ollama/server.ChatHandler (6 handlers)
[GIN-debug] GET    /                         --> github.com/jmorganca/ollama/server.(*Server).GenerateRoutes.func2 (5 handlers)
[GIN-debug] GET    /api/tags                 --> github.com/jmorganca/ollama/server.ListModelsHandler (5 handlers)
[GIN-debug] GET    /api/version              --> github.com/jmorganca/ollama/server.(*Server).GenerateRoutes.func3 (5 handlers)
[GIN-debug] HEAD   /                         --> github.com/jmorganca/ollama/server.(*Server).GenerateRoutes.func2 (5 handlers)
[GIN-debug] HEAD   /api/tags                 --> github.com/jmorganca/ollama/server.ListModelsHandler (5 handlers)
[GIN-debug] HEAD   /api/version              --> github.com/jmorganca/ollama/server.(*Server).GenerateRoutes.func3 (5 handlers)
time=2024-03-24T11:21:16.097-06:00 level=INFO source=routes.go:999 msg="Listening on 127.0.0.1:11434 (version 0.0.0)"
time=2024-03-24T11:21:16.097-06:00 level=INFO source=payload_common.go:106 msg="Extracting dynamic libraries..."
time=2024-03-24T11:21:16.178-06:00 level=INFO source=payload_common.go:145 msg="Dynamic LLM libraries [oneapi cpu_avx2 cpu_avx cpu]"
time=2024-03-24T11:21:16.178-06:00 level=INFO source=gpu.go:105 msg="Detecting GPU type"
time=2024-03-24T11:21:16.178-06:00 level=INFO source=gpu.go:285 msg="Searching for GPU management library libnvidia-ml.so"
time=2024-03-24T11:21:16.182-06:00 level=INFO source=gpu.go:331 msg="Discovered GPU libraries: []"
time=2024-03-24T11:21:16.182-06:00 level=INFO source=gpu.go:285 msg="Searching for GPU management library librocm_smi64.so"
time=2024-03-24T11:21:16.183-06:00 level=INFO source=gpu.go:331 msg="Discovered GPU libraries: []"
time=2024-03-24T11:21:16.183-06:00 level=INFO source=gpu.go:285 msg="Searching for GPU management library libze_intel_gpu.so"
time=2024-03-24T11:21:16.187-06:00 level=INFO source=gpu.go:331 msg="Discovered GPU libraries: [/usr/lib/x86_64-linux-gnu/libze_intel_gpu.so.1.3.27191.42]"
time=2024-03-24T11:21:16.209-06:00 level=INFO source=gpu.go:130 msg="Intel GPU detected"
time=2024-03-24T11:21:16.209-06:00 level=INFO source=cpu_common.go:11 msg="CPU has AVX2"
time=2024-03-24T11:21:16.210-06:00 level=INFO source=routes.go:1022 msg="no GPU detected"
```

Probalby due to the unified memory of the integrated GPU. I am not experienced with golang or project but a look into the routes.go file gives me the impression that the VRAM Check fails due to unified memory

```golang
1019         if runtime.GOOS == "linux" { // TODO - windows too
1020                 // check compatibility to log warnings
1021                 if _, err := gpu.CheckVRAM(); err != nil {
1022                         slog.Info(err.Error())
1023                 }
1024         }
```

Any suggestions what i could do to work around this?

**Comment by airMeng** [comment]
@semidark yes, I think there is some confusing in the logic here, we will refactor later to align with llama.cpp side.

**Comment by dhiltgen** [comment]
@zhewang1-intc I just got my hands on an Intel ARC GPU and my plan is to get it installed in a system, and try this PR out this week, then I'll do a code review for you.

File: Dockerfile
Comment (dhiltgen) [inline Dockerfile:58]: We recently cleaned up the package paths, so this is stale.  replace jmorganca with ollama.

**Review by dhiltgen — COMMENTED** [review]
I haven't looked through the code in depth yet, but when I tried to run this on my test system, something doesn't work correctly during discovery.
```
time=2024-04-01T22:05:13.407Z level=DEBUG source=gpu.go:332 msg="gpu management search paths: [/usr/lib/x86_64-linux-gnu/libze_intel_gpu.so* /usr/lib*/libze_intel_gpu.so* /home/daniel/libze_intel_gpu.so*]"
time=2024-04-01T22:05:13.408Z level=INFO source=gpu.go:360 msg="Discovered GPU libraries: [/usr/lib/x86_64-linux-gnu/libze_intel_gpu.so.1.3.28202.39]"
wiring Level-Zero management library functions in /usr/lib/x86_64-linux-gnu/libze_intel_gpu.so.1.3.28202.39
dlsym: zesInit
dlsym: zesDriverGet
dlsym: zesDeviceGet
dlsym: zesDeviceGetProperties
dlsym: zesDeviceEnumMemoryModules
dlsym: zesMemoryGetProperties
dlsym: zesMemoryGetState
zesInit err: 2013265921
time=2024-04-01T22:05:13.430Z level=INFO source=gpu.go:406 msg="Unable to load oneAPI management library /usr/lib/x86_64-linux-gnu/libze_intel_gpu.so.1.3.28202.39: oneapi vram init failure: 2013265921"
```

I was trying to emulate a user environment (not developer), so I only install the driver packages for ubuntu.

**Comment by zhewang1-intc** [comment]
> I haven't looked through the code in depth yet, but when I tried to run this on my test system, something doesn't work correctly during discovery.
> 
> ```
> time=2024-04-01T22:05:13.407Z level=DEBUG source=gpu.go:332 msg="gpu management search paths: [/usr/lib/x86_64-linux-gnu/libze_intel_gpu.so* /usr/lib*/libze_intel_gpu.so* /home/daniel/libze_intel_gpu.so*]"
> time=2024-04-01T22:05:13.408Z level=INFO source=gpu.go:360 msg="Discovered GPU libraries: [/usr/lib/x86_64-linux-gnu/libze_intel_gpu.so.1.3.28202.39]"
> wiring Level-Zero management library functions in /usr/lib/x86_64-linux-gnu/libze_intel_gpu.so.1.3.28202.39
> dlsym: zesInit
> dlsym: zesDriverGet
> dlsym: zesDeviceGet
> dlsym: zesDeviceGetProperties
> dlsym: zesDeviceEnumMemoryModules
> dlsym: zesMemoryGetProperties
> dlsym: zesMemoryGetState
> zesInit err: 2013265921
> time=2024-04-01T22:05:13.430Z level=INFO source=gpu.go:406 msg="Unable to load oneAPI management library /usr/lib/x86_64-linux-gnu/libze_intel_gpu.so.1.3.28202.39: oneapi vram init failure: 2013265921"
> ```
> 
> I was trying to emulate a user environment (not developer), so I only install the driver packages for ubuntu.

It seems that the err code `2013265921` indicate the `ZE_RESULT_ERROR_UNINITIALIZED` error in level-zero, which means the driver is not initialized.
Please make sure that the Intel GPU driver is installed correctly on your machine.
You can refer to the [README-sycl](https://github.com/ggerganov/llama.cpp/blob/master/README-sycl.md) in llama.cpp to setup the environment and make sure that the device check and examples in the README-sycl work normally. 
e.g.` ZES_ENABLE_SYSMAN=1 ./build/bin/main -m models/llama-2-7b.Q4_0.gguf -p "Building a website can be done in 10 simple steps:" -n 400 -e -ngl 33 -sm none -mg 0`

**Comment by airMeng** [comment]
Hi it shall work now after https://github.com/ggerganov/llama.cpp/pull/6435 merged

**Comment by zhewang1-intc** [comment]
@dhiltgen hi, could you pls take a review when you free so we can improve this pr?

**Comment by dhiltgen** [comment]
Since I'm having some difficulty successfully getting this running on my test system, let me suggest we break this into 2 pieces to reduce the rebase churn so we can make incremental progress.  Let's focus this PR on the base enablement with the `./gpu/*` updates and `./server/*` changes so that this can be built from source if you have the libraries on your host, but otherwise is dormant code.  Then create a new PR for the `./docs/*` and `./scripts/*` changes which we can merge later once things are well tested and solid across a broad set of distros with ~zero manual steps required by the user beyond installing the GPU Driver.  I'll take a review pass through this PR focused on the GPU and server changes.

File: llm/generate/gen_linux.sh
Comment (dhiltgen) [inline llm/generate/gen_linux.sh:221]: This feels like it may be problematic if the user doesn't have oneapi installed at the exact same location as the build system.  I think we'll eventually need to carry these libraries as dependency payloads like we do with cuda and rocm.  (assuming that's permitted)

As an incremental step before we expose in the official builds, this may be OK, although I see you are copying the libraries below.  Ideally we'd like to set this up so the user only needs the driver installed on the host, and we use the user-space library from our build to ensure things are linked properly.

File: llm/server.go
Comment (dhiltgen) [inline llm/server.go:230]: This doesn't seem necessary.   [This block](https://github.com/ollama/ollama/blob/main/llm/server.go#L260-L279) should cover it, and if you need additional deps wired up, use GpuInfo.DependencyPath.

File: llm/generate/gen_linux.sh
Comment (zhewang1-intc) [reply llm/generate/gen_linux.sh:221]: The libraries copied to the ${BUILD_DIR}/bin directory contain the essential oneAPI dependencies required for running the llama.cpp. There's no need for users to install oneAPI themselves, as the build system handles this dependency which means build system should install oneAPI.

When the ollama binary executes, these dependencies are extracted to a temporary location on the user's machine. Upon detecting an Intel GPU driver, we appends the path to this temporary directory (containing the oneAPI dependencies) to the LD_LIBRARY_PATH environment variable. This ensures that even if oneAPI isn't installed locally, the program can still locate the necessary libraries to function.

File: llm/server.go
Comment (zhewang1-intc) [reply llm/server.go]: yes, you are right, i remove it and ollama also works even if we didn't source the oneapi-related env script.

File: llm/generate/gen_linux.sh
Comment (dhiltgen) [reply llm/generate/gen_linux.sh:221]: In that case, I don't think this rpath is what we want, since that's setting a runtime setting for where ld can find the libraries at runtime and assumes you have installed the libs in the same location as the build system, but we likely want it to be a relative path based on `$ORIGIN` like we're doing with ROCm.  We can tidy this up in a follow up.

**Review by dhiltgen — APPROVED** [review]

**Comment by airMeng** [comment]
@dhiltgen Anything needing before merged?

**Comment by dhiltgen** [comment]
I'll merge this once we finalized the 0.1.39 release

**Comment by DocMAX** [comment]
```
time=2024-12-15T18:40:08.077Z level=INFO source=types.go:131 msg="inference compute" id=GPU-c54579ac-8f8e-f3a7-141b-fea32408ffd0 library=cuda variant=v12 compute=8.6 driver=12.7 name="NVIDIA GeForce RTX 3080" total="9.7 GiB" available="7.9 GiB"
time=2024-12-15T18:40:08.077Z level=INFO source=types.go:131 msg="inference compute" id=0 library=oneapi variant="" compute="" driver=0.0 name="Intel(R) Arc(TM) A380 Graphics" total="5.9 GiB" available="5.6 GiB"
time=2024-12-15T18:40:08.077Z level=INFO source=types.go:131 msg="inference compute" id=1 library=oneapi variant="" compute="" driver=0.0 name="Intel(R) Arc(TM) A310 LP Graphics" total="4.0 GiB" available="3.8 GiB"
```
Does this log mean my Intel ARCs are supported? Right now they are not used together with the NVIDIA GPU.

**Pull request merged**
